maven-extension/src/main/java/io/opentelemetry/maven/OtelExecutionListener.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package io.opentelemetry.maven;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.maven.semconv.MavenOtelSemanticAttributes;
+import java.util.Locale;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.apache.maven.execution.AbstractExecutionListener;
+import org.apache.maven.execution.ExecutionEvent;
+import org.apache.maven.execution.ExecutionListener;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.plugin.MojoExecution;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.component.annotations.Component;
+import org.codehaus.plexus.component.annotations.Requirement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Close the OpenTelemetry SDK (see {@link OpenTelemetrySdkService#dispose()} on the end of
+ * execution of the last project ({@link #projectSucceeded(ExecutionEvent)} and {@link
+ * #projectFailed(ExecutionEvent)}) rather than on the end of the Maven session {@link
+ * #sessionEnded(ExecutionEvent)} because OpenTelemetry and GRPC classes are unloaded by the Maven
+ * classloader before {@link #sessionEnded(ExecutionEvent)} causing {@link NoClassDefFoundError}
+ * messages in the logs.
+ */
+@Component(role = ExecutionListener.class, hint = "otel-execution-listener")
+public class OtelExecutionListener extends AbstractExecutionListener {
+
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  @Requirement private SpanRegistry spanRegistry;
+
+  @Requirement private OpenTelemetrySdkService openTelemetrySdkService;
+
+  /**
+   * Register in given {@link OtelExecutionListener} to the lifecycle of the given {@link
+   * MavenSession}
+   *
+   * @see org.apache.maven.execution.MavenExecutionRequest#setExecutionListener(ExecutionListener)
+   */
+  public static void registerOtelExecutionListener(
+      @Nonnull MavenSession session, @Nonnull OtelExecutionListener otelExecutionListener) {
+    @Nullable
+    ExecutionListener initialExecutionListener = session.getRequest().getExecutionListener();
+    if (initialExecutionListener instanceof ChainedExecutionListener) {
```

Comment:
IIUC, due to a limitation of maven's extensions, there isn't really a way to guarantee a different extension didn't overwrite our execution listener after we run. This is OK, we should add to the README if there's any POM ordering constraint.

Reply:
I didn't identify any constraint on the ordering of dependencies in the pom.xml. The order in which Maven loads `LifecycleParticipant`s is not specified, if this Otel lifecycle participant doesn't wrap all the extensions, then the measurement mays be slightly less accurate but I don't expect this to be an issue as it's uncommon in the Maven ecosystem to have extensions hooking in the lifecycle.